File: src/main/java/martian/minefactorial/datagen/MFDataGen.java
package martian.minefactorial.datagen;

import martian.minefactorial.datagen.client.MFBlockStateProvider;
import martian.minefactorial.datagen.client.MFItemModelProvider;
import net.minecraft.data.DataGenerator;
import net.minecraft.data.PackOutput;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.data.event.GatherDataEvent;

public final class MFDataGen {
	private MFDataGen() { }

	public static void onGatherData(final GatherDataEvent event) {
		if (event.includeClient()) {
			runClientGenerators(event);
		}

		if (event.includeServer()) {
			runServerGenerators(event);
		}
	}

	private static void runClientGenerators(final GatherDataEvent event) {
		DataGenerator dataGenerator = event.getGenerator();
		PackOutput packOutput = dataGenerator.getPackOutput();
		ExistingFileHelper efh = event.getExistingFileHelper();

		dataGenerator.addProvider(true, new MFBlockStateProvider(packOutput, efh));
		dataGenerator.addProvider(true, new MFItemModelProvider(packOutput, efh));
	}

	private static void runServerGenerators(final GatherDataEvent event) {
	}
}

File: src/data/kotlin/martian/minefactorial/datagen/MFDataGen.kt
package martian.minefactorial.datagen

import martian.dapper.api.add
import martian.dapper.api.addProviders
import martian.minefactorial.Minefactorial
import martian.minefactorial.datagen.client.MFBlockStateProvider
import martian.minefactorial.datagen.client.MFFusionMetadataProvider
import martian.minefactorial.datagen.client.MFFusionModelProvider
import martian.minefactorial.datagen.client.MFItemModelProvider
import martian.minefactorial.datagen.server.*
import net.minecraft.data.PackOutput
import net.minecraft.resources.ResourceLocation
import net.neoforged.fml.common.Mod
import net.neoforged.neoforge.data.event.GatherDataEvent
import thedarkcolour.kotlinforforge.neoforge.forge.MOD_BUS

@Mod(Minefactorial.MODID + "_data")
object MFDataGen {
	init {
		MOD_BUS.addListener(::onGatherData)
	}

	fun onGatherData(event: GatherDataEvent) {
		event.addProviders(
			client = {
				it.add(MFBlockStateProvider(event))
				it.add(MFItemModelProvider(event))

				val fusionPackOutput = PackOutput(event.generator.packOutput.outputFolder.resolve("fusion"))
				println(fusionPackOutput.outputFolder.toString())
				it.add(MFFusionModelProvider(fusionPackOutput))
				it.add(MFFusionMetadataProvider(fusionPackOutput))
			},
			server = {
				it.add(MFLootTableProvider(event))
				it.add(MFBlockTagsProvider(event))
				it.add(MFFluidTagsProvider(event))
				it.add(MFItemTagsProvider(event))
				it.add(MFRecipeProvider(event))
			}
		)
	}
}

val String.id: ResourceLocation get() = Minefactorial.id(this)
